### ISSUE TITLE
Option to enable/disable autograd.profiler, move Chrome trace processing to C++

### ISSUE THREAD

**Issue by heiner** [pytorch/pytorch#19420]
## 🚀 Feature
The context manager `torch.autograd.profiler.profile()` enables profiling by calling `torch.autograd._enable_profiler()` and `torch.autograd._disable_profiler()`.

Enabling/disabling this profiling on the fly should be fully supported. Right now exporting chrome traces as json breaks when calling _enable and _disable directly while training happens in another thread. (The `pop()` call in https://github.com/pytorch/pytorch/blob/master/torch/autograd/profiler.py#L465 can then happen on an empty list.)

While we're at it, building the Chrome tracing json file in Python has significant overhead and it seems reasonable to move that code into C++.

It also seems reasonable to export a gzipped json file, as Chrome can deal with those directly.

(It might also be worth investigating whether running an http server alongside PyTorch which displays tracing information on demand is something we could support. Other popular ML frameworks do that.)

## Motivation

In situations in which (1) there's many op executions and (2) the profiled code needs a certain burn-in time, the json file resulting from calling `export_chrome_trace()` can get many GB large. Building these large json files in Python takes a long time, and Chrome runs OOM when you try to load them.

Manually enabling/disabling profiling while other threads run PyTorch ops also breaks building the json traces as described above.

## Pitch

```python
torch.autograd.profiler.enable()
time.sleep(10)
prof = torch.autograd.profiler.disable()
prof.export_chrome_trace("tracefile.json.gz")  # Not in Python; fast.  
```

## Alternatives

None what. so. ever.

**Comment by ezyang** [comment]
We will support porting the trace processing to C++ (although, if you are exporting too many events, you will probably choke the trace viewer. It's better to trace less.)

In triage meeting, we weren't convinced by the justification for enable/disable. We think that many of the use-cases can be addressed by simply writing your tracing instead--so really this should be a documentation improvement. In particular, the suggestion is that you shouldn't trace over multiple iterations of the loop, instead you should trace one iteration (and disable profiling for the first few loops to avoid warmup). This can definitely be documented better. Would you like to help review this doc improvement when it's ready?

**Comment by heiner** [comment]
I appreciate this might not be a major concern in most applications. However, the suggestion to just not "trace over multiple iterations of the loop" makes some rather strong assumptions on what the application looks like, including the assumption of having a single loop.

For an example of where these assumptions don't hold, consider the case of a multi-threaded asynchronous system, e.g. in recent policy-gradient RL algorithms. Those systems tend to have a "burn-in" phase that isn't super interesting to profile, but that phase might in fact already generate more profiling data than the tools here can deal with.

**Comment by ezyang** [comment]
Yeah, I think it would help us a lot if you could give an example where it's not easy to restructure the code so that you trace only one iteration.

**Comment by v0dro** [comment]
So from what I understand of this issue, enabling and disabling profiler is not much of a priority, but porting the chrome tracing functionality to C++ is. Am I correct?

If yes, I'd like to take up the issue of porting the tracing to C++. Will send a sample proposal of my approach soon.

**Comment by ezyang** [comment]
@v0dro That looks like what this issue says!

**Comment by ezyang** [comment]
Some notes on the implementation:

1. I'm not too familiar with this part of the code, so make sure you verify things I say by reading the code. Adam Paszke is the original author of the code in question if you can get in touch with him.
2. It certainly seems plausible that creation of the Python data structures which are subsequently dumped by json is expensive. Note that this is actually a two step process: we record events in C++, we convert them to Python, and then we do a second conversion to put them into the chrome tracing format
3. It's not enough to only port export_chrome_trace; look at profile.__exit__, we also have to parse_cpu_trace on the returned records
4. There's no json encoder in C++ standard library. How exactly the JSON serialization will be done is a big concern. Before you start coding there should be some proposal about how this should be done. As best as I can tell, there is no precedent for JSON serialization from C++ in PyTorch today
5. Profiler code is bound to Python in torch/csrc/autograd/init.cpp; look there to see how to expose your C++ functionality

**Comment by v0dro** [comment]
## Basics

Current benchmarks are as follows:
``` python
    import time
    import torch
    
    x = torch.ones(2, 2)
    with torch.autograd.profiler.profile() as prof: 
      for _ in range(10000):
        x * x
    
    times = []
    for i in range(50):
      start = time.time()
      prof.export_chrome_trace("trace.json")
      stop = time.time()
      times.append(stop-start)
    
    
    print(sum(times)/len(times))
    # => 9.271759657859802
```

Going to the `chrome://tracing` url within google chrome allows you to see a function
call trace of the code.


<a id="orgf72eaca"></a>

## Software mini-proposal


<a id="org59f49f2"></a>

### Goal

Port the `exportc_chrome_tracing` python function to C++ for greater speed.


### Problem

Currently the chrome tracing recording and export is quite slow.

### Profiling current implementation

``` python
    import torch
    import cProfile, pstats
    from io import StringIO
    
    x = torch.ones(2, 2)
    
    # events are recorded here in a C++ codebase in the prof object.
    pr = cProfile.Profile()
    pr.enable()
    
    with torch.autograd.profiler.profile() as prof: 
      for _ in range(10000):
        x * x
    prof.export_chrome_trace("trace.json")
    
    pr.disable()
    
    s = StringIO()
    sortby = 'cumulative'
    ps = pstats.Stats(pr, stream=s).sort_stats(sortby)
    ps.print_stats()
    print(s.getvalue())
```

The profiler output is as follows:

            1    0.002    0.002    0.488    0.488 /profiler.py:287(export_chrome_trace)
            1    0.031    0.031    0.466    0.466 /profiler.py:94(export_chrome_trace)
            1    0.083    0.083    0.429    0.429 /json/__init__.py:120(dump)
       300002    0.053    0.000    0.296    0.000 /json/encoder.py:412(_iterencode)
       300002    0.068    0.000    0.243    0.000 /json/encoder.py:277(_iterencode_list)
    320000/300000    0.102    0.000    0.168    0.000 /json/encoder.py:333(_iterencode_dict)
            1    0.000    0.000    0.115    0.115 /profiler.py:259(__exit__)
            1    0.083    0.083    0.103    0.103 /profiler.py:641(parse_cpu_trace)

Most of the time is taken by the `export_chrome_trace` function. Reducing this time
would make things much faster.

### Lit survey

The profiling module exists in the C++ interface as `RecordProfile`. However it has a different
API that requires one to create a CPP block and then call the `RecordProfile` constructor
inside this block in order to profile the block. How does RecordProfile output JSON files?

This previous PR focuses on outputting the profiler in the form of an output stream
that be seen by a user using something like cout. It records these events using enableProfiler
and then gets the events in a vector with disableProfiler. These are then put in the outstream
in the process<sub>events</sub> function that is called inside the desctructor.

However the `RecordProfile` class does not have any provision to output the events
through the Python API.

### Possible solution

The function `export_chrome_function` exists within `torch/autograd/profiler.py`.
It is basically called from the `profiler` class but ultimately a call to the 
same-named function within the `EventList` class is made. Porting this function
to C++ is the main issue.

Current workflow ::

-   **Make events in C++:** This is done inside `autograd._disable_profiler`. It returns a list of C++ objects bound inside a python list.
-   **`parse_cpu_time`:** convert this list of C++ objects into python `FunctionEvent` objects.
-   **`export_chrome_trace`:** converts `FunctionEvent` objects into dict format and dumps this structure as JSON.

Instead of this conversion into python list ->  FunctionEvent -> JSON maybe it should
be possible to directly utilize the C++ `Event` class and output the chrome tracing
format JSON.

The `RecordProfile` class already has functionality for generating an event trace 
within C++. This can be used as inspiration to build a chain between the event
generation and output as a JSON and do everything directly within C++.


<a id="org7113024"></a>

### Implementation overview

The `autograd._disable_profiler` function that called withing `profile.__exit__` already
calls the `disableProfiler` C++ function that returns a list of C++ objects. However
the next thing in the `__exit__` method is the `parse_cpu_trace` method that converts
these events into Python objects.

The gist of the plan is to save the output of `autograd._disable_profiler` in an instance
variable `self.records` and lazily call `parse_cpu_trace` in whatever methods requires
it. The call should not be done in the `__exit__` method. Once the user calls 
`export_chrome_trace` on the `profile` object, we call a C++ function that is directly
passed the `self.records` variable and generates a JSON dump from this object directly.
In theory this should speed up the execution.

The effect of this on other methods within the `profile` class will be that there
will be some if checks within the functions like `key_averages` that will generate
the `function_events` when the method is called and save it in `self.function_events`.

### Implementation details

The implementation will involve two parts:

1.  Converting `Event` objects into JSON-like format.
2.  Dumping the `Event` objects as JSON.

The nice thing about every single object in the `export_chrome_trace` function
looks like this:

``` python
    [{"name": "mul", "ph": "X", "ts": 35316.272, "dur": 2.2410000000018044, "tid": 0, "pid": "CPU functions", "args": {}}]
```

Therefore the format that each of these follows is:

    - top level array object
      - dict object
        - name : string
        - ph : string
        - ts : float
        - dur : float
        - tid : int
        - pid : string
        - args : dict
      - another dict object
        - same thing as above

Thus there is no nested serialization or something crazy of the sort happening inside,
which should make building a tiny C++ JSON dumping class a not-so-tough job. This
new C++ JSON dumping class will start off as a minimal implementation that is meant
to only support dumping trace events (with focus on extensibility). Since it will
be available to the entire pytorch codebase, someone who might need a JSON dumping
class later can extend it to support nested objects etc. as per their need.

**Comment by v0dro** [comment]
@ezyang can you please have a look at my mini-proposal? I'll improve on the details and post more stuff soon.

**Comment by ezyang** [comment]
> This previous PR focuses on outputting the profiler in the form of an output stream
> that be seen by a user using something like cout.

I think you're referring to #16580 from @zdevito 

> The RecordProfile class already has functionality for generating an event trace within C++. 

OK, so are you saying we should take RecordProfile and increase its functionality so that it completely subsumes the Python side version of the logic, or are you going to make another class that is like RecordProfile, but separate?

> The gist of the plan is to save the output of autograd._disable_profiler in an instance
> variable self.records and lazily call parse_cpu_trace in whatever methods requires
> it. The call should not be done in the __exit__ method.

I don't know what this means.

> Thus there is no nested serialization or something crazy of the sort happening inside,
> which should make building a tiny C++ JSON dumping class a not-so-tough job. 

You could also do it YOLO similar to RecordFunction is doing it:

```
  TORCH_CHECK(start, "could not find start?");
  std::vector<Event*> stack;
  out_ << "[\n";
  bool first = true;
  for(Event* e : events) {
    if(e->kind() == "push") {
      stack.push_back(e);
    } else if(e->kind() == "pop") {
      if(!first) {
        out_ << ",\n";
      }
      first = false;
      Event* e_start = stack.back();
      stack.pop_back();
      jit::TemplateEnv env;
      env.s("name", e_start->name());
      env.d("ts", start->cpu_elapsed_us(*e_start));
      env.d("dur", e_start->cpu_elapsed_us(*e));
      env.d("tid", e_start->thread_id());
      out_ << event_template.format(env);
    }
  }
  out_ << "]\n";
}
```

**Comment by ezyang** [comment]
I think you're OK to start implementing.

**Comment by v0dro** [comment]
Can this issue be considered closed?
